src/openvic-simulation/GameManager.cpp

```diff
@@ -0,0 +1,132 @@
+#include "GameManager.hpp"
+
+#include "openvic-simulation/utility/Logger.hpp"
+
+using namespace OpenVic;
+
+GameManager::GameManager(state_updated_func_t state_updated_callback)
+	: clock { [this]() { tick(); }, [this]() { update_state(); } },
+	  state_updated { state_updated_callback } {}
+
+void GameManager::set_needs_update() {
+	needs_update = true;
+}
+
+void GameManager::update_state() {
+	if (needs_update) {
+		Logger::info("Update: ", today);
+		map.update_state(today);
+		if (state_updated) state_updated();
+		needs_update = false;
+	}
```

Comment:
Probably should prefer:
```suggestion
	if (needs_update) return;
	Logger::info("Update: ", today);
	map.update_state(today);
	if (state_updated) state_updated();
	needs_update = false;
```

Reply:
I assume you meant `if (!needs_update) return;`, and yeah your way is better, the overwhelming majority of calls to `update_state` will be while `needs_update` is false.

Thinking about similar cases throughout the code, is it worth using C++20's `[[likely]]` and `[[unlikely]]`, or should I just let the compiler handle it, writing the likely case first as a hint where possible?

Reply:
`[[likely]]` and `[[unlikely]]` are optimizations for when the compiler can't see whats going on, the only reason to use it honestly is when you can know that the compiler doesn't actually know how to predict the branches well, so you try to tell it early how it should predict, iirc msvc may still completely ignore this because of course it does, it kinda requires benchmarks to know if this optimization is necessary, because if you're wrong you just costed the branch predictor immensely.